Add Audio.MultiStageEnvelope shard for flexible envelope generation

sinkingsugar opened the pull request:
Implements a multi-stage envelope generator inspired by DX7's EG system:
- Unlimited attack/decay stages via Stages parameter (Float3 sequences)
- Per-stage curve control: logarithmic (-1), linear (0), exponential (1)
- SustainIndex for hold point during gate-on
- Separate Release stages for gate-off behavior
- Loop support with configurable LoopStart index
- Multi-channel SIMD processing (NEON/SSE batch of 4 channels)

Stage format: @f3(level, time, curve) where:
- level: target value (0.0-1.0)
- time: duration in seconds
- curve: interpolation shape (-1 to 1)

Includes comprehensive tests and audio demos.


<!--
Before submitting your PR, please review the following checklist:

- [ ] **DO NOT** submit a PR without having discussed the change first in a related issue. If none exist, create one first.
- [ ] **CONSIDER** adding a unit test if your PR resolves an issue.
- [ ] **DO** keep pull requests small so they can be easily reviewed.
- [ ] **DO** make sure all unit tests pass.
- [ ] **DO** make sure not to introduce any new compiler warnings.
- [ ] **AVOID** breaking the continuous integration build.
- [ ] **AVOID** making significant changes to the overall architecture.
-->

sinkingsugar committed:
Add Audio.MultiStageEnvelope shard for flexible envelope generation

Implements a multi-stage envelope generator inspired by DX7's EG system:
- Unlimited attack/decay stages via Stages parameter (Float3 sequences)
- Per-stage curve control: logarithmic (-1), linear (0), exponential (1)
- SustainIndex for hold point during gate-on
- Separate Release stages for gate-off behavior
- Loop support with configurable LoopStart index
- Multi-channel SIMD processing (NEON/SSE batch of 4 channels)

Stage format: @f3(level, time, curve) where:
- level: target value (0.0-1.0)
- time: duration in seconds
- curve: interpolation shape (-1 to 1)

Includes comprehensive tests and audio demos.

sinkingsugar committed:
Fix PR review issues for Audio.MultiStageEnvelope

- Add validation for LoopStart index (clamp to 0 if out of bounds)
- Add validation for SustainIndex (set to -1 if out of bounds)
- Initialize totalStageSamples to 1 before checking empty stages
- Remove unused SIMD curve functions (applyCurve_NEON, applyCurve_SSE)
- Simplify channel batch processing (remove SYNTH_HAS_NEON/SSE guards)
- Improve curve documentation with power function range details

sinkingsugar committed:
Improve Audio.MultiStageEnvelope based on PR review

- Add proper gate edge detection for retrigger from any phase
  (matches analog synth behavior - retrigger restarts from current level)
- Add prevGate tracking in ChannelState for rising/falling edge detection
- Call cacheStages() in activate() for dynamic parameter support
- Update help text: remove false SIMD claim, document retrigger behavior

sinkingsugar committed:
Fix out-of-bounds access when stages change dynamically

When Stages or Release parameters are reduced while envelope is running
at a later stage, state.currentStage could exceed the new cache size.

Added bounds check before accessing stage arrays:
- Attack phase: if currentStage >= _stagesCache.size(), transition to Sustain
- Release phase: if currentStage >= _releaseCache.size(), transition to Idle

This ensures safe behavior when users modify envelope parameters in real-time.

sinkingsugar commented:
> ## Code Review: Audio.MultiStageEnvelope
> 
> I've reviewed the implementation of the multi-stage envelope generator. The code is generally well-structured and follows project patterns. Below are my findings:
> 
> ---
> 
> ### ✅ **STRENGTHS**
> 
> 1. **Architecture**: Clean state machine design with clear phase transitions (Idle → Attack → Sustain → Release)
> 2. **Safety**: Proper bounds checking and validation throughout (sustain index, loop start, buffer sizes)
> 3. **Dynamic parameters**: Re-caching stages each frame (line 1291) allows runtime parameter changes
> 4. **Retrigger handling**: Smooth retriggering from current level prevents discontinuities (lines 1138-1149)
> 5. **Comprehensive testing**: Test suite covers basic functionality, curves, looping, multi-channel, and VCA usage
> 
> ---
> 
> ### 🔴 **CRITICAL ISSUES**
> 
> None found. The implementation is robust for realistic usage scenarios.
> 
> ---
> 
> ### 🟡 **MEDIUM ISSUES**
> 
> #### 1. **Loop discontinuity with non-matching levels** (Line 1200-1206)
> **Severity**: MEDIUM  
> **Confidence**: HIGH
> 
> When looping, if the final stage level doesn't match the loop start level, there will be an audible click/discontinuity:
> 
> ```cpp
> if (_looping && _loopStartIdx < _stagesCache.size()) {
>   state.currentStage = _loopStartIdx;
>   state.startLevel = state.level;  // Current level from stage N
>   // But next stage targets _stagesCache[_loopStartIdx].level
>   // If these don't match → instant jump
> ```
> 
> **Example**: 
> - Stages: [@f3(0.3, 0.02, 0.0), @f3(1.0, 0.02, 0.0), @f3(0.3, 0.02, 0.0)]
> - Loop: true, LoopStart: 0
> - At end of stage 2: level = 0.3
> - Loop to stage 0: targets level 0.3
> - **No problem in this case**
> 
> But with:
> - Stages: [@f3(0.5, 0.02, 0.0), @f3(1.0, 0.02, 0.0), @f3(0.3, 0.02, 0.0)]
> - Loop: true, LoopStart: 1
> - At end: level = 0.3
> - Loop to stage 1: targets 1.0
> - **Instant jump from 0.3 → 1.0 base** (though smoothed by curve)
> 
> **Fix**: Document this behavior in help text or add a transition stage automatically.
> 
> ---
> 
> #### 2. **Time = 0 creates instant jumps** (Lines 1147-1148, 1158-1159, 1205-1206, 1214-1215, 1255-1256)
> **Severity**: MEDIUM  
> **Confidence**: HIGH
> 
> Stages with `time: 0.0` result in `totalStageSamples = 1`, causing near-instant level changes. While this might be intentional for instant attacks, it could cause clicks if used mid-envelope.
> 
> ```cpp
> state.totalStageSamples = std::max(uint64_t(1), uint64_t(std::round(_stagesCache[0].time * sampleRate)));
> ```
> 
> With `time = 0.0` at 44100 Hz:
> - `std::round(0.0 * 44100) = 0`
> - `std::max(1, 0) = 1`
> - Jump occurs in 1 sample (0.023ms at 44.1kHz)
> 
> **Realistic concern**: Users creating complex envelopes might accidentally use 0.0 and wonder why they hear clicks.
> 
> **Fix options**:
> 1. Document minimum recommended time (e.g., 0.001 seconds)
> 2. Clamp minimum time to small value (e.g., `std::max(uint64_t(2), ...)`)
> 3. Add warning in help text
> 
> ---
> 
> ### 🟢 **LOW PRIORITY / STYLE**
> 
> #### 1. **Dynamic recaching performance** (Line 1291)
> `cacheStages()` is called every activate() to support dynamic parameters. For static envelopes, this is unnecessary work. Consider:
> - Dirty flag pattern (only recache when parameters change)
> - Trade-off: Added complexity vs. marginal CPU savings
> 
> **Verdict**: Current approach is fine unless profiling shows issues.
> 
> ---
> 
> #### 2. **Curve function numerical edge cases** (Lines 1113-1126)
> `applyCurve()` with extreme curve values near ±1:
> - `curve = 1.0`: exponent = 4.0 (fine)
> - `curve = -1.0`: uses `1 - (1-t)^4` (fine)
> - `t = 0 or 1`: `std::pow` handles correctly
> 
> No issues found - the math is bounded and safe.
> 
> ---
> 
> #### 3. **Comment accuracy** (Line 1110)
> Comment says "curve=-1 → t^(-2) equivalent" but code uses:
> ```cpp
> return 1.0f - std::pow(1.0f - t, 1.0f - curve * 3.0f);
> // For curve = -1: 1 - (1-t)^4
> ```
> 
> This is `1 - (1-t)^4`, not `t^{-2}`. The comment is misleading (though the behavior is correct for "fast start").
> 
> **Fix**: Update comment to match actual formula.
> 
> ---
> 
> ### 📋 **SUMMARY**
> 
> | Severity | Count | Actionable |
> |----------|-------|------------|
> | Critical | 0     | -          |
> | High     | 0     | -          |
> | Medium   | 2     | Yes        |
> | Low      | 3     | Optional   |
> 
> **Recommendation**: Address loop discontinuity documentation and consider time=0 handling before merging. The implementation is otherwise solid and production-ready.
> 
> **Test coverage**: Excellent - tests cover all major features including edge cases like looping, multi-channel, and VCA usage.

@claude can you address those medium issues at least please. They seem legit

github-actions[bot] committed:
Fix medium-priority issues in Audio.MultiStageEnvelope

- Enforce minimum stage duration of ~2ms to prevent clicks from time=0
- Add documentation about minimum stage times and loop level matching
- Fix misleading comment about curve math (curve=-1 is 1-(1-t)^4)
- Add clarifying comment about loop discontinuity prevention

Co-authored-by: Giovanni Petrantoni <sinkingsugar@users.noreply.github.com>

f3 commented:
@claude - I know you are a bot, but be careful with your PR review comments. Make sure that examples containing code snippets, like the one below, do not inadvertently include references that @-mention GitHub usernames. Otherwise they will be notified by mistake 😁

> 
> **Example**: 
> - Stages: [@f3(0.3, 0.02, 0.0), @f3(1.0, 0.02, 0.0), @f3(0.3, 0.02, 0.0)]
> - Loop: true, LoopStart: 0
> - At end of stage 2: level = 0.3
> - Loop to stage 0: targets level 0.3
> - **No problem in this case**
> 
> But with:
> - Stages: [@f3(0.5, 0.02, 0.0), @f3(1.0, 0.02, 0.0), @f3(0.3, 0.02, 0.0)]
> - Loop: true, LoopStart: 1
> - At end: level = 0.3
> - Loop to stage 1: targets 1.0
> - **Instant jump from 0.3 → 1.0 base** (though smoothed by curve)
>

sinkingsugar commented:
> @claude - I know you are a bot, but be careful with your PR review comments. Make sure that examples containing code snippets, like the one below, do not inadvertently include references that @-mention GitHub usernames. Otherwise they will be notified by mistake 😁
> 
> > 
> > **Example**: 
> > - Stages: [@f3(0.3, 0.02, 0.0), @f3(1.0, 0.02, 0.0), @f3(0.3, 0.02, 0.0)]
> > - Loop: true, LoopStart: 0
> > - At end of stage 2: level = 0.3
> > - Loop to stage 0: targets level 0.3
> > - **No problem in this case**
> > 
> > But with:
> > - Stages: [@f3(0.5, 0.02, 0.0), @f3(1.0, 0.02, 0.0), @f3(0.3, 0.02, 0.0)]
> > - Loop: true, LoopStart: 1
> > - At end: level = 0.3
> > - Loop to stage 1: targets 1.0
> > - **Instant jump from 0.3 → 1.0 base** (though smoothed by curve)
> > 
> 

Apologies @f3 but @f3 is a language keyword. This is a broader issue with GitHub. Thanks for understanding.

sinkingsugar commented:
> ## Code Review: Audio.MultiStageEnvelope
> 
> Reviewed the implementation of the multi-stage envelope generator. The code is well-structured and implements a flexible envelope system inspired by DX7. Here are my findings:
> 
> ---
> 
> ### ✅ Overall Assessment
> 
> The implementation is **solid** with good defensive programming patterns. The code handles edge cases well, supports dynamic parameter changes, and processes multi-channel audio efficiently.
> 
> ---
> 
> ### 🔴 CRITICAL Issues
> 
> **None found.** The code handles realistic edge cases appropriately.
> 
> ---
> 
> ### 🟡 MEDIUM Issues
> 
> #### 1. Loop discontinuity risk with mismatched levels (synth.cpp:1205-1211)
> **Location:** Lines 1205-1211
> 
> When looping, if the final stage level doesn't match the loop start stage level, there will be an audible discontinuity (click/pop). The code tries to prevent this by using `state.startLevel = state.level`, but this only smooths the *transition into* the next stage - it doesn't prevent the discontinuity when the final level (e.g., 0.5) jumps to the loop start target level (e.g., 1.0).
> 
> ```cpp
> if (_looping && _loopStartIdx < _stagesCache.size()) {
>   // Loop back - start from current level to prevent discontinuity
>   state.currentStage = _loopStartIdx;
>   state.startLevel = state.level;  // This smooths the transition IN, not the jump
>   state.stageSamples = 0;
>   state.totalStageSamples = ...
> }
> ```
> 
> **Impact:** Audible clicks when looping with mismatched levels.
> 
> **Recommendation:** Consider adding a brief crossfade stage when looping back, or document more prominently that loop start and end levels must match. The help text mentions this (line 1039) but users may miss it.
> 
> ---
> 
> #### 2. Potential integer overflow in stage sample calculation (synth.cpp:1150-1151)
> **Location:** Lines 1150-1151, 1162-1163, 1210-1211, 1220-1221, 1262-1263
> 
> ```cpp
> state.totalStageSamples = std::max(
>     uint64_t(sampleRate / 500), 
>     uint64_t(std::round(_stagesCache[0].time * sampleRate))
> );
> ```
> 
> **Issue:** If `_stagesCache[0].time` is very large (e.g., 10000.0 seconds at 44100 Hz = 441M samples), the multiplication `time * sampleRate` happens in `float` before conversion to `uint64_t`. Floats lose precision above ~16M, potentially causing incorrect stage durations.
> 
> **Impact:** Very long stage times (>6 minutes) may have inaccurate durations.
> 
> **Likelihood:** LOW - typical musical envelopes use times < 10 seconds.
> 
> **Fix:** Cast before multiplication: `uint64_t(stage.time) * sampleRate` or use double precision for intermediate calculation.
> 
> ---
> 
> ### 🟢 LOW Issues / Style Suggestions
> 
> #### 3. Redundant caching on every activate() call (synth.cpp:1298)
> **Location:** Line 1298
> 
> ```cpp
> // Re-cache stages each frame to support dynamic parameter changes
> cacheStages();
> ```
> 
> **Observation:** This is called on every audio buffer, which is fine for supporting dynamic parameter changes. However, if parameters rarely change in practice, this could be optimized with a dirty flag.
> 
> **Impact:** Minor CPU overhead (likely negligible compared to audio processing).
> 
> **Recommendation:** Consider adding a dirty flag if profiling shows this is a bottleneck. Current approach favors simplicity and correctness, which is reasonable.
> 
> ---
> 
> #### 4. Comment mismatch in applyCurve() (synth.cpp:1126)
> **Location:** Line 1126
> 
> ```cpp
> // Inverse of exponential: 1 - (1-t)^(1 + |curve|*3)
> return 1.0f - std::pow(1.0f - t, 1.0f - curve * 3.0f);
> ```
> 
> **Issue:** Comment says `(1 + |curve|*3)` but code uses `(1.0f - curve * 3.0f)`. Since `curve < 0` here, `1.0f - curve * 3.0f` equals `1.0f - (negative) * 3.0f` = `1.0f + |curve| * 3.0f`, so the code is correct but the comment could be clearer.
> 
> **Recommendation:** Update comment to: `1 - (1-t)^(1 - curve*3)` (matching the code exactly) to avoid confusion.
> 
> ---
> 
> ### ✅ Good Practices Observed
> 
> 1. **Minimum time enforcement** (lines 1150, 1162, 1210, 1220, 1262): The ~2ms minimum (`sampleRate / 500`) prevents clicks from instant jumps.
> 2. **Bounds validation** (lines 1097-1104): Sustain and loop indices are validated and clamped.
> 3. **Retrigger handling** (lines 1140-1165): Retriggering from any phase works smoothly by starting from current level.
> 4. **Multi-channel processing** (lines 1305-1323): Batch processing for cache efficiency.
> 5. **Dynamic parameter support** (line 1298): Re-caching enables live parameter changes.
> 6. **Comprehensive tests**: Test coverage is excellent (basic, curves, complex, loop, VCA, multi-channel).
> 
> ---
> 
> ### 📝 Summary
> 
> This is a **well-implemented feature** with good attention to audio-specific concerns (click prevention, smooth retriggering). The issues found are:
> - **Medium #1**: Looping discontinuity (already documented in help text)
> - **Medium #2**: Very long stage times may have precision issues (edge case)
> - **Low #3-4**: Minor optimization opportunity and comment clarity
> 
> **Recommendation:** ✅ **Approve** - The medium issues are acceptable given their low likelihood in real usage. Consider addressing #2 if you expect users to create very long envelopes (>6 minutes).

@claude pls assess those comments, they seem valid.

sinkingsugar commented:
> @claude - I know you are a bot, but be careful with your PR review comments. Make sure that examples containing code snippets, like the one below, do not inadvertently include references that @-mention GitHub usernames. Otherwise they will be notified by mistake 😁
> 
> > 
> > **Example**: 
> > - Stages: [@f3(0.3, 0.02, 0.0), @f3(1.0, 0.02, 0.0), @f3(0.3, 0.02, 0.0)]
> > - Loop: true, LoopStart: 0
> > - At end of stage 2: level = 0.3
> > - Loop to stage 0: targets level 0.3
> > - **No problem in this case**
> > 
> > But with:
> > - Stages: [@f3(0.5, 0.02, 0.0), @f3(1.0, 0.02, 0.0), @f3(0.3, 0.02, 0.0)]
> > - Loop: true, LoopStart: 1
> > - At end: level = 0.3
> > - Loop to stage 1: targets 1.0
> > - **Instant jump from 0.3 → 1.0 base** (though smoothed by curve)
> > 
> 

@claude add this remark by user f3 to the two GitHub CI actions involving you please

sinkingsugar committed:
Improve claude CI

sinkingsugar committed:
Improve test coverage for Audio.MultiStageEnvelope

Add 4 new test wires to exercise uncovered code paths:

- test-audio-multistage-gate-transitions: Tests gate on→off transitions,
  sustain entry, multi-stage release phase, and stage completion logic
- test-audio-multistage-loop-completion: Tests loop restart when reaching
  end of stages, covering loop back from different LoopStart indices
- test-audio-multistage-dynamic-stages: Tests empty stages array and
  empty release handling for graceful fallback behavior
- test-audio-multistage-retrigger: Tests retrigger from various phases
  with multiple gate transitions

These tests cover the previously untested lines including:
- Stage completion and next stage transition
- Sustain index matching during attack
- Loop back when end of stages reached
- Release stage transitions
- Empty stages/release graceful handling
